src/lib.rs

```diff
@@ -438,338 +507,521 @@ impl TsunamiBuilder {
+                            setup_fn(&mut ssh).map(move |_| ssh).then(move |r| {
+                                let r = r.context(format!(
+                                    "setup procedure for {} machine failed",
+                                    name
+                                ));
+
+                                if r.is_err() {
+                                    error!(log, "setup for {} machine failed", name);
+                                }
+
+                                r.map(move |ssh| (machine, ssh))
+                            })
+                        })
+                        .map(move |(mut machine, ssh)| {
+                            info!(log, "finished setting up {} instance", name; "dns" => &machine.public_dns);
+                            machine.ssh = Some(ssh);
+                            (name, machine)
                         })
-                            .map_err(Error::from),
-                    )
-                } else {
-                    Either::B(wait_for_instances)
-                }.map(move |(instances, id_to_name, ok)| {
-                    (instances, ok, id_to_name, setup_fns, usernames, pk)
                 })
             })
+            .collect();
+
+        let future = futures::future::join_all(setups)
             .map_err(Error::from)
-            .and_then(
-                move |(instances, everything_ok, id_to_name, setup_fns, usernames, private_key)| {
-                    let mut term_instances = instances.clone();
-
-                    // or_else
-                    let at_the_end = move |e| {
-                        use std::mem;
-
-                        // 5. terminate all instances
-                        if !term_instances.is_empty() {
-                            debug!(log, "terminating instances");
-                            let mut termination_req =
-                                rusoto_ec2::TerminateInstancesRequest::default();
-                            termination_req.instance_ids =
-                                mem::replace(&mut term_instances, Vec::new());
-
-                            Either::A(ec2.terminate_instances(&termination_req).map_err(
-                                move |term_e| {
-                                    warn!(
-                                        log,
-                                        "failed to terminate tsunami instances: {:?}", term_e
-                                    );
-                                    // TODO: include both errors?
-                                    e
-                                },
-                            ))
-                        } else {
-                            Either::B(futures::future::err(e))
-                        }.map(|_| ())
-                            .map_err(Error::from)
+            .and_then(move |machines| {
+                let machines: HashMap<&str, _> = machines.into_iter().fold(
```

Comment:
You likely won't need the explicit type annotation here if you make the change to `name` I suggested above.